### PR TITLE
Move structure validation to separate structure validator

### PR DESCRIPTION
**Pull request by jhugman**
Relates to [ EXP-3938](https://mozilla-hub.atlassian.net/browse/EXP-3938).

According to [The Big O of Code Reviews](https://www.egorand.dev/the-big-o-of-code-reviews/), this is a O(_n_) change.

This is moving quite a lot of code out `intermediate_representation.rs` into a new `structure/validator.rs` file.

The changes are fairly mechanical, but they added up. The tests also changed and moved, so this is a larger O(_n_) sized.

This has a requirement on landing #5947 and rebasing before this will build.

### Pull Request checklist ###

- **Breaking changes**:  This PR follows our [breaking change policy](https://github.com/mozilla/application-services/blob/main/docs/howtos/breaking-changes.md)
  - [ ] This PR follows the breaking change policy:
     - This PR has no breaking API changes, or
     - There are corresponding PRs for our consumer applications that resolve the breaking changes and have been approved
- [ ] **Quality**: This PR builds and tests run cleanly
  - Note:
    - For changes that need extra cross-platform testing, consider adding `[ci full]` to the PR title.
    - If this pull request includes a breaking change, consider [cutting a new release](https://github.com/mozilla/application-services/blob/main/docs/howtos/cut-a-new-release.md) after merging.
- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Changelog**: This PR includes a changelog entry in [CHANGELOG.md](../CHANGELOG.md) or an explanation of why it does not need one
  - Any breaking changes to Swift or Kotlin binding APIs are noted explicitly
- [ ] **Dependencies**: This PR follows our [dependency management guidelines](https://github.com/mozilla/application-services/blob/main/docs/dependency-management.md)
  - Any new dependencies are accompanied by a summary of the due dilligence applied in selecting them.

[Branch builds](https://github.com/mozilla/application-services/blob/main/docs/howtos/branch-builds.md): add `[firefox-android: branch-name]` to the PR title.

### CODE REVIEW

File: components/support/nimbus-fml/src/structure/validator.rs [hunk 1]
```diff
@@ -0,0 +1,485 @@
...
+
+            // Check the types exist for this property.
+            let path = format!("objects/{obj_nm}/{prop_nm}");
+            self.validate_type_ref(&path, &prop.typ)?;
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn validate_feature_def(&self, feature_def: &FeatureDef) -> Result<()> {
```
Comment (jhugman) [inline components/support/nimbus-fml/src/structure/validator.rs:50]: This method is the largest non-mechanical change: here we're aggregating a lot of disparate methods which were all iterating over `feature_def.props`, and put them all in one loop.

File: components/support/nimbus-fml/src/structure/validator.rs [hunk 2]
```diff
@@ -0,0 +1,485 @@
...
+            "some_def",
+            "test doc",
+            vec![PropDef::new(
+                "prop name",
+                TypeRef::Enum("EnumDoesntExist".into()),
+                json!(null),
+            )],
+            false,
+        );
+        validator.validate_feature_def(&fm).expect_err(
```
Comment (jhugman) [inline components/support/nimbus-fml/src/structure/validator.rs:218]: All these tests used to be of the form:

```
let fm = get_feature_manifest(…)

fm.validate_something()
```

I've left most of the variable names in tact.

**Review by jhugman — COMMENTED** [review]
Self review.

**Review by eliserichards — APPROVED** [review]
Perfect! 👏

**Pull request merged**